File: csharp/Resources/AuditResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment audit Wave 2 operations
    /// </summary>
    public class AuditResource
    {
        private readonly AuthdogClient _client;

        public AuditResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListLogsAsync(string tenantId, string environmentId, object? query = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/audit/logs",
                query: AuthdogClient.QueryFrom(query));

        public JObject ListLogs(string tenantId, string environmentId, object? query = null) =>
            ListLogsAsync(tenantId, environmentId, query).GetAwaiter().GetResult();

        public Task<JObject> EventMetadataAsync(string tenantId, string environmentId, object? query = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/audit/event-metadata",
                query: AuthdogClient.QueryFrom(query));

        public JObject EventMetadata(string tenantId, string environmentId, object? query = null) =>
            EventMetadataAsync(tenantId, environmentId, query).GetAwaiter().GetResult();

        public Task<JObject> EventTypesAsync(string tenantId, string environmentId, object? query = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/audit/event-types",
                query: AuthdogClient.QueryFrom(query));

        public JObject EventTypes(string tenantId, string environmentId, object? query = null) =>
            EventTypesAsync(tenantId, environmentId, query).GetAwaiter().GetResult();

        public Task<JObject> EventTypesCatalogAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/audit/event-types/catalog");

        public JObject EventTypesCatalog(string tenantId, string environmentId) =>
            EventTypesCatalogAsync(tenantId, environmentId).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/AuditResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * Environment audit log operations.
 */
public final class AuditResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an audit helper.
     * @param clientParam Authdog client
     */
    public AuditResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List environment audit logs.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return logs envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listLogs(final String tenantId,
                             final String environmentId)
            throws AuthenticationException, ApiException {
        return listLogs(tenantId, environmentId, null);
    }

    /**
     * List environment audit logs with caller query parameters.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param params optional query parameters
     * @return logs envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listLogs(final String tenantId,
                             final String environmentId,
                             final Map<String, String> params)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/audit/logs",
                null, params, JsonNode.class);
    }

    /**
     * Get environment audit event metadata.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return metadata envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode eventMetadata(final String tenantId,
                                  final String environmentId)
            throws AuthenticationException, ApiException {
        return eventMetadata(tenantId, environmentId, null);
    }

    /**
     * Get environment audit event metadata with query parameters.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param params optional query parameters
     * @return metadata envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode eventMetadata(final String tenantId,
                                  final String environmentId,
                                  final Map<String, String> params)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/audit/event-metadata",
                null, params, JsonNode.class);
    }

    /**
     * Get grouped environment audit event types.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return event types envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode eventTypes(final String tenantId,
                               final String environmentId)
            throws AuthenticationException, ApiException {
        return eventTypes(tenantId, environmentId, null);
    }

    /**
     * Get grouped environment audit event types with query
     * parameters.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param params optional query parameters
     * @return event types envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode eventTypes(final String tenantId,
                               final String environmentId,
                               final Map<String, String> params)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/audit/event-types",
                null, params, JsonNode.class);
    }

    /**
     * Get the audit event type catalog.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return catalog envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode eventTypesCatalog(final String tenantId,
                                      final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/audit/event-types/catalog",
                null, null, JsonNode.class);
    }
}
